src/StateMaps/PDEConstrainedFunctionals.jl

```diff
@@ -428,3 +428,342 @@ function Base.show(io::IO,::MIME"text/plain",f::AbstractPDEConstrainedFunctional
+
+  # Unpack
+  j = val[1]
+  c = Vector{eltype(val)}()
+  grad = first(_grad)
+  dJ = grad[1]
+  dC = Vector{eltype(grad)}();
+
+  # Analytic derivative
+  function _compute_dF!(dF,analytic_dF!::Function)
+    analytic_dF!(dF,get_free_dof_values(φh))
+    nothing
+  end
+  function _compute_dF!(dF,analytic_dF!::Nothing)
+    nothing
+  end
+  _compute_dF!(dJ,pcf.analytic_dJ)
+
+  return j,c,dJ,dC
+end
+
+function get_state(m::CustomPDEConstrainedFunctionals)
+  @warn """
+    For CustomPDEConstrainedFunctionals, get_state only returns the StateMap used in
+    the constructor for the CustomPDEConstrainedFunctionals. This means that uh in
+    ```
+    for (it,uh,φh) in optimiser
+      ...
+    end
+    ```
+    will correspond to this particular state map.
+
+    There may be cases where you have multiple StateMaps inside a map φ_to_jc. If
+    this is the case, you should get your states directly from your StateMaps in
+    the driver script.
+
+    This functionality may change in future.
+  """ maxlog=1
+  get_state(m.state_map)
+end
+get_state_map(m::CustomPDEConstrainedFunctionals) = m.state_map
+
+function evaluate_functionals!(pcf::CustomPDEConstrainedFunctionals,φh)
+  φ = get_free_dof_values(φh)
+  return evaluate_functionals!(pcf,φ)
+end
+
+function evaluate_functionals!(pcf::CustomPDEConstrainedFunctionals{N},φ::AbstractVector) where N
+  φ_to_jc =  pcf.φ_to_jc
+  val = φ_to_jc(φ)
+  j = val[1]
+  c = val[2:end];
+  return j,c
+end
+
+function evaluate_functionals!(pcf::CustomPDEConstrainedFunctionals{0},φ::AbstractVector)
+  φ_to_jc =  pcf.φ_to_jc
+  val = φ_to_jc(φ)
+  j = val[1]
+  c = Vector{eltype(val)}();
+  return j,c
+end
+
+########## Zygote + Unfitted ##########
+"""
+    struct CustomEmbeddedPDEConstrainedFunctionals{N,A} <: AbstractPDEConstrainedFunctionals{N}
+
+A version of `CustomPDEConstrainedFunctionals` that has an `embedded_collection` to
+allow the `state_map` to be updated given new FE spaces for the forward problem.
+This is currently required for unfitted methods.
+"""
+struct CustomEmbeddedPDEConstrainedFunctionals{N,A} <:  AbstractPDEConstrainedFunctionals{N}
+  φ_to_jc :: Function
+  analytic_dJ
+  analytic_dC
+  embedded_collection :: EmbeddedCollection
+    @doc"""
+        CustomEmbeddedPDEConstrainedFunctionals(
+          φ_to_jc :: Function,
+          num_constraints,
+          embedded_collection :: EmbeddedCollection;
+          analytic_dJ = nothing,
+          analytic_dC = fill(nothing,num_constraints)
+        )
+
+    Create an instance of `CustomEmbeddedPDEConstrainedFunctionals`. Here,
+    `num_constraints` specifies the number of constraints.
+    """
+    function CustomEmbeddedPDEConstrainedFunctionals(
+      φ_to_jc :: Function,
+      num_constraints,
+      embedded_collection :: EmbeddedCollection;
+      analytic_dJ = nothing,
+      analytic_dC = fill(nothing,num_constraints)
+    )
+
+    @check Set((:state_map,:J,:C)) == keys(embedded_collection.objects) """

```

Comment:
Why do we need these exact keys? Is it to seperate the 2 embedded collections? I feel we lose generality with this way of seperating them. It might not be a good idea to be forced to have a :J and a :C (or a state_map either). Reason being is that your objective (which we commonly think of as J) may not be the `StateParamMap` that we define in the `EmbeddedCollection` like it is in the common case. We might have J instead a function of the `StateParamMap` which we compose inside `φ_to_jc` or J1 + J2 + J3 etc. Same goes for C. C might not be something that requires embedded machinery, for example. Also, It would also make sense to use all this machinery for a `φ_to_jc` that doesnt even have a state_map inside. Conversely we might have many state_maps but neither fit this template.

Reply:
Another alternative, is to allow the user to not specify anything - then simply return nothing in the 'states' part of the optimiser iterator (get_states -> nothing).